File: C++/Graphs/06_EvaluateDivision.cpp
#include <bits/stdc++.h>
using namespace std;
unordered_map<string, vector<pair<string, double>>> l;

void dfs(string src, string dest, double tmp, double &sum, bool &found, map<string, bool> &vis)
{
    vis[src] = true;
    if (src == dest)
    {
        found = true;
        sum = tmp;
        return;
    }

    for (auto i : l[src])
    {
        if (vis.find(i.first) == vis.end())
            dfs(i.first, dest, tmp * i.second, sum, found, vis);
    }
}

vector<double> calcEquation(vector<vector<string>> &eq, vector<double> &values, vector<vector<string>> &query)
{

    vector<double> ans;
    int n = eq.size();

    for (int i = 0; i < n; i++)
    {
        double w = values[i];
        double w1 = 1 / w;

        l[eq[i][0]].push_back({eq[i][1], w});
        l[eq[i][1]].push_back({eq[i][0], w1});
    }
    for (int i = 0; i < query.size(); i++)
    {

        map<string, bool> vis;

        if (l.find(query[i][0]) == l.end())
        {
            ans.push_back(-1.0);
            continue;
        }
        else
        {
            bool found = false;
            double sum = 0.0;
            dfs(query[i][0], query[i][1], 1.0, sum, found, vis);
            if (!found)
                ans.push_back(-1.0);
            else
                ans.push_back(sum);
        }
    }
    return ans;
}

File: Java/Graphs/06_EvaluateDivision.java
class Solution {
    public class Edge{
        String v;
        double w;
        Edge(String v,double w){
            this.v = v;
            this.w = w;
        }
    }
    double res = 1.0;
    public boolean dfs(HashMap<String,ArrayList<Edge>> map, String i,String j,HashSet<String> vis){
        if(i.equals(j)){
            // System.out.println("Yes");
            return true;
        }
        vis.add(i);
        ArrayList<Edge> al = map.get(i);
        for(Edge e: al){
            // System.out.println(e.v + ", " + e.w);
            if(!vis.contains(e.v)){
                res *= e.w;
                if(dfs(map,e.v,j,vis)) return true;
                res /= e.w;
            }
        }
        return false;
    }
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        HashMap<String,ArrayList<Edge>> map = new HashMap<>();
        int n = values.length;
        for(int i=0;i<n;i++){
            String u = equations.get(i).get(0);
            String v = equations.get(i).get(1);
            double w = values[i];
            
            if(!map.containsKey(u)){
                map.put(u, new ArrayList<>());
            }
            if(!map.containsKey(v)){
                map.put(v, new ArrayList<>());
            }
            map.get(u).add(new Edge(v,w));
            map.get(v).add(new Edge(u,(double)1/w));
        }
        double[] ans = new double[queries.size()];
        int idx = 0;
        for(List<String> q: queries){
            String src = q.get(0);
            String dest = q.get(1);
            
            if(!map.containsKey(src) || !map.containsKey(dest)){
                ans[idx++] = -1.00000;
                continue;
            } 
            HashSet<String> vis = new HashSet<>();
            
            ans[idx++] = (dfs(map,src,dest,vis)) ?  res : -1.0;
            res = 1.00000;
        }
        return ans;
    }
}